File: custom_dir/custom_actions/time_check.py
# -*- coding: UTF-8 -*-
import json
import datetime
from maa.context import Context
from maa.custom_action import CustomAction
from maa.agent.agent_server import AgentServer


@AgentServer.custom_action("TimeCheck")
class TimeCheck(CustomAction):
    def run(self,
            context: Context,
            argv: CustomAction.RunArg) -> bool:
        """
        自定义动作：检查当前系统时间是否在指定范围内（支持时间 + 日期规则）

        该动作用于判断“当前系统时间”是否满足指定的时间区间，
        并可选地结合日期规则（具体日期 / 每月几号 / 每周星期几）进行限制。
        若未提供日期规则，则默认对所有日期生效（完全兼容旧逻辑）。

        :param argv: custom_action_param，JSON 格式字符串，示例：

                     {
                        "start": "23:00",        // 开始时间 (HH:MM 或 HH:MM:SS)
                        "end": "06:00",          // 结束时间 (HH:MM 或 HH:MM:SS)
                        "mode": "in",            // 可选，默认 "in"
                                                  // "in"  : 在时间段内返回 True
                                                  // "out" : 在时间段外返回 True

                        "date_rule": {           // 【可选】日期规则，不传表示不限制日期
                            "type": "week",      // 规则类型，见下方说明
                            "value": [1, 2, 3]   // 规则值
                        }
                     }

        【date_rule.type 支持类型说明】

        - "date"      : 指定某一天
                        value 示例："2025-01-20"

        - "dates"     : 指定多个具体日期
                        value 示例：["2025-01-20", "2025-01-25"]

        - "month_day" : 每月的某几号
                        value 示例：[1, 15, 28]

        - "week"      : 每周星期几（ISO 标准）
                        1=周一 ... 7=周日
                        value 示例：[1, 2, 3, 4, 5]

        - "none" 或 不传 date_rule
                      : 不进行日期限制，任意日期均可

        :param context: 执行上下文（由框架传入）

        :return:
            - 当 mode = "in"  时：
                当前时间 + 日期规则 均满足条件 → 返回 True
                否则 → 返回 False

            - 当 mode = "out" 时：
                当前时间或日期规则不满足 → 返回 True
                否则 → 返回 False
        """

        print("开始执行自定义动作：时间 + 日期范围检查")

        # ========== 1. 参数解析 ==========
        try:
            params = json.loads(argv.custom_action_param)
            if not isinstance(params, dict):
                print("参数格式错误：期望 JSON 对象")
                return False

            start_str = params.get("start")
            end_str = params.get("end")
            mode = params.get("mode", "in").lower()
            date_rule = params.get("date_rule")  # 新增日期规则

            if not start_str or not end_str:
                print("错误：必须提供 start 和 end 时间")
                return False

        except json.JSONDecodeError as e:
            print(f"参数JSON解析失败: {e}")
            return False

        # ========== 2. 当前时间 ==========
        now_dt = datetime.datetime.now()
        now_time = now_dt.time()
        today = now_dt.date()

        # ========== 3. 时间解析 ==========
        try:
            def parse_time(t_str):
                try:
                    return datetime.datetime.strptime(t_str, "%H:%M").time()
                except ValueError:
                    return datetime.datetime.strptime(t_str, "%H:%M:%S").time()

            start_time = parse_time(start_str)
            end_time = parse_time(end_str)

            print(f"当前时间: {now_dt.strftime('%Y-%m-%d %H:%M:%S')}")
            print(f"时间范围: {start_time} - {end_time}, 模式: {mode}")

        except ValueError as e:
            print(f"时间格式错误: {e}")
            return False

        # ========== 4. 日期规则判断 ==========
        def check_date_rule(rule) -> bool:
            """不传规则默认 True"""
            if not rule:
                return True

            rule_type = rule.get("type")
            value = rule.get("value")

            if rule_type == "date":
                target = datetime.datetime.strptime(value, "%Y-%m-%d").date()
                return today == target

            elif rule_type == "dates":
                targets = [
                    datetime.datetime.strptime(d, "%Y-%m-%d").date()
                    for d in value
                ]
                return today in targets

            elif rule_type == "month_day":
                # 每月几号
                return today.day in value

            elif rule_type == "week":
                # Monday=1 ... Sunday=7
                return today.isoweekday() in value

            elif rule_type == "none":
                return True

            else:
                print(f"未知 date_rule.type: {rule_type}")
                return False

        date_ok = check_date_rule(date_rule)
        print(f"日期规则检查结果: {date_ok}")

        if not date_ok:
            return False if mode == "in" else True

        # ========== 5. 时间范围判断 ==========
        is_in_range = False

        if start_time < end_time:
            # 同一天
            if start_time <= now_time <= end_time:
                is_in_range = True
        else:
            # 跨天
            if now_time >= start_time or now_time <= end_time:
                is_in_range = True

        # ========== 6. 模式判断 ==========
        if mode == "in":
            result = is_in_range
        elif mode == "out":
            result = not is_in_range
        else:
            print(f"未知模式 '{mode}'")
            return False

        print(f"最终结果: {result}")
        return result

    def stop(self) -> None:
        pass

File: agent/actions/time_check/time_check.go
package time_check

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type TimeCheck struct{}

type TimeCheckParams struct {
	Start    string `json:"start"`
	End      string `json:"end"`
	Mode     string `json:"mode"`
	DateRule struct {
		Type  string      `json:"type"`
		Value interface{} `json:"value"`
	} `json:"date_rule"`
}

// Run 检查当前系统时间是否在指定范围内
func (a *TimeCheck) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	fmt.Println("开始执行自定义动作：时间 + 日期范围检查")

	params := TimeCheckParams{Mode: "in"}
	if arg.CustomActionParam != "" {
		if err := json.Unmarshal([]byte(arg.CustomActionParam), &params); err != nil {
			fmt.Printf("参数JSON解析失败: %v\n", err)
			return false
		}
	}

	if params.Start == "" || params.End == "" {
		fmt.Println("错误：必须提供 start 和 end 时间")
		return false
	}

	now := time.Now()
	nowTime := now.Format("15:04:05")
	today := now.Format("2006-01-02")

	fmt.Printf("当前时间: %s %s\n", today, nowTime)
	fmt.Printf("时间范围: %s - %s, 模式: %s\n", params.Start, params.End, params.Mode)

	// 解析时间
	startTime, err := time.Parse("15:04", params.Start)
	if err != nil {
		startTime, err = time.Parse("15:04:05", params.Start)
		if err != nil {
			fmt.Printf("时间格式错误: %v\n", err)
			return false
		}
	}

	endTime, err := time.Parse("15:04", params.End)
	if err != nil {
		endTime, err = time.Parse("15:04:05", params.End)
		if err != nil {
			fmt.Printf("时间格式错误: %v\n", err)
			return false
		}
	}

	// 检查日期规则
	dateOK := a.checkDateRule(params.DateRule, now)
	fmt.Printf("日期规则检查结果: %v\n", dateOK)

	if !dateOK {
		if params.Mode == "in" {
			return false
		}
		return true
	}

	// 解析当前时间
	nowTimeParsed, _ := time.Parse("15:04:05", nowTime)

	// 时间范围判断
	isInRange := false
	if startTime.Before(endTime) {
		// 同一天
		if (nowTimeParsed.After(startTime) || nowTimeParsed.Equal(startTime)) &&
			(nowTimeParsed.Before(endTime) || nowTimeParsed.Equal(endTime)) {
			isInRange = true
		}
	} else {
		// 跨天
		if nowTimeParsed.After(startTime) || nowTimeParsed.Before(endTime) {
			isInRange = true
		}
	}

	// 模式判断
	var result bool
	if params.Mode == "in" {
		result = isInRange
	} else if params.Mode == "out" {
		result = !isInRange
	} else {
		fmt.Printf("未知模式 '%s'\n", params.Mode)
		return false
	}

	fmt.Printf("最终结果: %v\n", result)
	return result
}

func (a *TimeCheck) checkDateRule(rule struct {
	Type  string      `json:"type"`
	Value interface{} `json:"value"`
}, now time.Time) bool {
	if rule.Type == "" || rule.Type == "none" {
		return true
	}

	today := now.Format("2006-01-02")

	switch rule.Type {
	case "date":
		if str, ok := rule.Value.(string); ok {
			return today == str
		}
	case "dates":
		if arr, ok := rule.Value.([]interface{}); ok {
			for _, d := range arr {
				if str, ok := d.(string); ok && str == today {
					return true
				}
			}
			return false
		}
	case "month_day":
		if arr, ok := rule.Value.([]interface{}); ok {
			day := now.Day()
			for _, d := range arr {
				if v, ok := d.(float64); ok && int(v) == day {
					return true
				}
			}
			return false
		}
	case "week":
		if arr, ok := rule.Value.([]interface{}); ok {
			weekday := int(now.Weekday())
			if weekday == 0 {
				weekday = 7 // 将周日从0改为7
			}
			for _, d := range arr {
				if v, ok := d.(float64); ok && int(v) == weekday {
					return true
				}
			}
			return false
		}
	}

	return true
}